File: 05- Hashmap/C#/3- Word Pattern.cs
public class Solution {
        public bool WordPattern(string pattern, string s) {
            Dictionary<char , string> mp = new Dictionary<char, string>(); 
            Dictionary<string , bool> valid = new Dictionary<string, bool>();
            string[] str = s.Split(' ');

            char[] pat = pattern.ToCharArray() ; 
            if (str.Length != pattern.Length)
                return false ;

            
            for (int i = 0 ; i < pat.Length ; i++){
                if (!mp.ContainsKey(pat[i]) && !valid.ContainsKey(str[i])){
                    mp[pat[i]] = str[i] ; 
                    valid[str[i]] = true ;
                }
                else {
                    if (!mp.ContainsKey(pat[i]))
                        return false ; 
                    if (mp[pat[i]] != str[i])
                        return false;
                }
            }
            return true ; 
        }
    }

File: 05- Hashmap/C++/3- Word Pattern.cpp
class Solution {
    public:
        bool wordPattern(string pattern, string s) {
            unordered_map<char , string> mp ;
            unordered_map<string , bool> valid ;
    
            vector<string> str;
            stringstream ss(s);
            string word;
    
            while (ss >> word) {
                str.push_back(word);
            }
            
            if (pattern.size() != str.size())
                return false ;
    
            for (int i = 0 ; i < pattern.size() ; i++) {
                if (mp.find(pattern[i]) == mp.end() && valid.find(str[i]) == valid.end()) {
                    mp[pattern[i]] = str[i] ;
                    valid[str[i]] = true ;
                }
                else {
                    if (mp.find(pattern[i]) == mp.end())
                        return false ;
                    if (mp[pattern[i]] != str[i])
                        return false ;
                }
            }
    
            return true;
        }
    };